homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,403 @@
+            return True
+        if identifier == dev.name:
+            return True
+        return any([service.identifier == identifier for service in dev.services])
+
+    def _host_filter():
+        try:
+            return [ip_address(identifier)]
+        except ValueError:
+            return None
+
+    if cache:
+        matches = [atv for atv in cache if _filter_device(atv)]
+        if matches:
+            return cache, matches[0]
+
+    for hosts in [_host_filter(), None]:
+        scan_result = await scan(loop, timeout=3, hosts=hosts)
+        matches = [atv for atv in scan_result if _filter_device(atv)]
+
+        if matches:
+            return scan_result, matches[0]
+
+    return scan_result, None
+
+
+def is_valid_credentials(credentials):
+    """Verify that credentials are valid for establishing a connection."""
+    return (
+        credentials.get(Protocol.MRP.value) is not None
+        or credentials.get(Protocol.DMAP.value) is not None
+    )
+
+
+class AppleTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_invalid_credentials(self, info):
+        """Handle initial step when updating invalid credentials."""
+        await self.async_set_unique_id(info.get(CONF_IDENTIFIER))
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["title_placeholders"] = {"name": info.get(CONF_NAME)}
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["identifier"] = self.unique_id
+        return await self.async_step_reconfigure()
+
+    async def async_step_reconfigure(self, user_input=None):
+        """Inform user that reconfiguration is about to start."""
+        if user_input is not None:
+            return await self.async_find_device_wrapper(
+                self.async_begin_pairing, allow_exist=True
+            )
+
+        return self.async_show_form(step_id="reconfigure")
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        # Be helpful to the user and look for devices
+        if self.scan_result is None:
+            self.scan_result, _ = await device_scan(None, self.hass.loop)
+
+        errors = {}
+        default_suggestion = self._prefill_identifier()
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_IDENTIFIER])
```

Comment:
We normally check for existing entries with same unique id and abort if there are any after setting unique id.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

Reply:
I will look into this. I believe I'm doing this wrong here.